### ISSUE TITLE
BayesOptSearch in tune

### ISSUE THREAD

**Issue by rkarn** [ray-project/ray#13962]
### What is the problem?

I would like to point to the Ray Tune python implementation of Bayesian search. At this code given at https://github.com/ray-project/ray/blob/master/python/ray/tune/suggest/bayesopt.py . 

The Bayesian optimization doesn't seems to have been integrated fully. As per the documentation given at https://github.com/fmfn/BayesianOptimization,   the objective function name should be supplied to the "BayesianOptimization" class. Some thing like this:
optimizer = BayesianOptimization(
    f=black_box_function,
    pbounds=pbounds,
    random_state=1,
)

But in the implementation given at https://github.com/ray-project/ray/blob/master/python/ray/tune/suggest/bayesopt.py, the optimization is setup like this:
self.optimizer = byo.BayesianOptimization(
            f=None,
            pbounds=self._space,
            verbose=self._verbose,
            random_state=self._random_state)  .

Why is " f=None". How to insert the neural network model objective function instead of "None" in the above implementation in Raytune?  
Currently, the same value of parameter variable is used in all the iterations of each trial. It should be different as shown in the https://github.com/fmfn/BayesianOptimization. The algosithm should search heavily in the red spot area as displayed in the animation of that github page. For that to happen, the value of parameter should change during the iterations of each trial.

**Comment by richardliaw** [comment]
hi @rkarn , I think the implementation is safe. Can you provide a repro example of the constant parameter?

**Comment by rkarn** [comment]
Hi @richardliaw , Thank you for responding. 

Consider the example given at https://docs.ray.io/en/master/tune/examples/bayesopt_example.html . I run this example. It has three parameter variables "steps, width, and height" and most of the trials are run for 100 iterations. The problem is,  the value of  variables "steps, width, and height" at iteration 1 is same as that of iteration 100. So, the parameter variable are not tuned in the respective trial. Screenshot attached ( taken from result.json file for one of the trial).

![image](https://user-images.githubusercontent.com/20768904/107149782-10911b00-6974-11eb-96be-db0e61b3d2fa.png)


As per Bayesian search, the search process is random in the given search space but after few iterations,  the algorithm should search heavily in the near optimal region (also called hot spot area) where the probability of finding the  optimal parameter value is maximum. That's how the Bayesian algorithm is explained in https://github.com/fmfn/BayesianOptimization. Please refer to the animation diagram at their github page.  

To obey such behavior, the BayesOptSearch should either tune the parameter variables during the trial or should create new search points based on past results.

**Comment by richardliaw** [comment]
@rkarn if you look closely, you're showing a screen shot of the *same* trial being evaluated multiple iterations.

I ran the script and got the following:

```
+-------------------------+------------+---------------------+----------+----------+----------+--------+------------------+--------------+-----------------+
| Trial name              | status     | loc                 |   height |    width |     loss |   iter |   total time (s) |   iterations |   neg_mean_loss |
|-------------------------+------------+---------------------+----------+----------+----------+--------+------------------+--------------+-----------------|
| easy_objective_cfbcb7d0 | RUNNING    | 192.168.1.115:32489 | -95.8831 | 19.3982  | -9.45446 |     39 |      4.06674     |           38 |         9.45446 |
| easy_objective_cb4c8e0a | TERMINATED |                     | -25.092  | 19.0143  | -2.42641 |     64 |      6.7521      |           63 |         2.42641 |
| easy_objective_cb4f3aec | TERMINATED |                     |  46.3988 | 11.9732  |  4.72354 |    100 |     10.5823      |           99 |        -4.72354 |
| easy_objective_cb503d0c | TERMINATED |                     | -68.7963 |  3.11989 | -6.56602 |    100 |     10.6083      |           99 |         6.56602 |
| easy_objective_cb512898 | TERMINATED |                     | -88.3833 | 17.3235  | -8.78036 |    100 |     10.6076      |           99 |         8.78036 |
| easy_objective_cf8af3e4 | TERMINATED |                     |  20.223  | 14.1615  | 12.0223  |      1 |      6.00815e-05 |            0 |       -12.0223  |
| easy_objective_d1d29d50 | TERMINATED |                     |  66.4885 |  4.24678 | 16.6489  |      1 |      7.72476e-05 |            0 |       -16.6489  |
| easy_objective_d1d76600 | TERMINATED |                     | -63.635  |  3.66809 |  3.6365  |      1 |      7.98702e-05 |            0 |        -3.6365  |
| easy_objective_d1d9bfb8 | TERMINATED |                     | -39.1516 | 10.4951  |  6.08484 |      1 |      0.00011611  |            0 |        -6.08484 |
| easy_objective_d20cfa7c | TERMINATED |                     | -13.611  |  5.82458 |  8.6389  |      1 |      7.79629e-05 |            0 |        -8.6389  |
```

as you can see, each trial runs a different number of iterations and has a different configuration.

**Comment by rkarn** [comment]
Hi @richardliaw , I agree with the screenshot you provided. My question was different.  I run that example with "n_samples=10" on a 4 core machine. Each trial run on 1 core resulting in 4 concurrent trials run in parallel.  To validate that BayesOpt is working effectively, can you please guide me in the following:

![image](https://user-images.githubusercontent.com/20768904/107205458-55ba5900-6a17-11eb-92ea-1cedfda7ab28.png)

I generated the above screenshot from the "analysis dataframe".  In BayesOptSearch,  parameter variable search points of the next trials are estimated based on the performance of previous trials.  To prove it, can I say that the parameter variable for trial # 5,6,7,8 is estimated based on the result obtained from trials 1 to 4? Similarly, parameter variable for trial # 9,10 is estimated based on the result obtained from trials 5 to 8? If that's the case then can I prove that trials  5,6,7,8 are generated after completion of trials 1 to 4? I don't see any significant difference between the timestamp of all 10 trials.

**Comment by richardliaw** [comment]
Can you set your concurrencylimiter to max_concurrent=1?

**Comment by rkarn** [comment]
I did that. I used the below script to display timestamps in a human-readable format. Further, I also captured the last modified timestamp of each trial folder (to denote each trial completion time).  

![image](https://user-images.githubusercontent.com/20768904/107625042-2d2f9a80-6c75-11eb-8982-04ed092e1638.png)

My question still remains as below:
1. How can we prove that the search point for next trail is estimated based on Bayesian search? In the Bayesian Search documentation, the API is used as :
optimizer = BayesianOptimization(
f=black_box_function,
pbounds=pbounds,
random_state=1,
)

But in your implementation you have used it as:
self.optimizer = byo.BayesianOptimization(
f=None,
pbounds=self._space,
verbose=self._verbose,
random_state=self._random_state) .

***Why is " f=None"? Can you please show me the location where you have called the Bayesian search API and run it?***

**Comment by richardliaw** [comment]
@rkarn please see https://github.com/fmfn/BayesianOptimization/blob/master/examples/advanced-tour.ipynb for the documentation on why `f=None` is fine and how we use BayesianOptimization in Ray Tune.

**Comment by hy-chen** [comment]
@rkarn Hi there, I think you should set the `random_search_steps` parameter for bayesopt. That defines how many points to randomly sample before bayesian optimization is used. those points will be used as the ground for later sampling. This is to help prevent overfitting to a local region. By default its 10, i think that might be why your experiment seems like totally random (as your total sample size is 10).

`    def __init__(self,
                 space: Optional[Dict] = None,
                 metric: Optional[str] = None,
                 mode: Optional[str] = None,
                 points_to_evaluate: Optional[List[Dict]] = None,
                 utility_kwargs: Optional[Dict] = None,
                 random_state: int = 42,
                 random_search_steps: int = 10,
                 verbose: int = 0,
                 patience: int = 5,
                 skip_duplicate: bool = True,
                 analysis: Optional[ExperimentAnalysis] = None,
                 max_concurrent: Optional[int] = None,
                 use_early_stopped_trials: Optional[bool] = None):`
